rtrlib/lib/vector.c

```diff
@@ -0,0 +1,154 @@
+	if (data_field == NULL) {
+		return -1;
+	}
+
+	struct aspa_vector *vector = (struct aspa_vector *)malloc(sizeof(struct aspa_vector));
+
+	if (vector == NULL) {
+		return -1;
+	}
+
+	vector->capacity = DEFAULT_INITIAL_SIZE;
+	vector->size = 0;
+	vector->data = data_field;
+
+	*vector_pointer = vector;
+
+	return 0;
+}
+
+int aspa_vector_delete(struct aspa_vector *vector)
+{
+	// checking if one of the two pointers is potentially invalid
+	if (vector == NULL || vector->data == NULL) {
+		return -1;
+	}
+
+	// freeing the data
+	free(vector->data);
+	// freeing the object itself
+	free(vector);
+
+	return 0;
+}
+
+int aspa_vector_reallocate(struct aspa_vector *vector)
+{
+	// by how many more elements this vector can hold after the relocation
+	const size_t SIZE_INCREASE_EXPONENTIAL = 2;
+
+	// allocation the new chunk of memory
+	struct aspas_record *new_data_field = (struct aspas_record *)malloc(
+		sizeof(struct aspas_record) * vector->capacity * SIZE_INCREASE_EXPONENTIAL);
+
+	// malloc failed so returning an error
+	if (new_data_field == NULL) {
+		return -1;
+	}
+
+	// copying the data from the old location to the new one
+	memcpy(new_data_field, vector->data, vector->capacity * sizeof(struct aspas_record));
+	vector->data = new_data_field;
+	vector->capacity = powl(vector->capacity, SIZE_INCREASE_EXPONENTIAL);
+
+	return 0;
+}
+
+void aspa_vector_sort(struct aspa_vector *vector, struct aspas_record record)
+{
+	size_t j = vector->size;
+
+	// moving data along
+	while (j > 0 && vector->data[j - 1].customer_as_number > record.customer_as_number) {
+		vector->data[j] = vector->data[j - 1];
+		j -= 1;
+	}
+
+	vector->data[j] = record;
+}
+
+int aspa_vector_insert(struct aspa_vector *vector, struct aspas_record record)
+{
+	// check if this element will fit into the vector
+	if (vector->size + 1 > vector->capacity) {
+		if (aspa_vector_reallocate(vector) == -1) {
+			return -1;
+		}
+	}
+
+	// insert the element at the correct place
+	aspa_vector_sort(vector, record);
+	vector->size += 1;
+
+	return 0;
+}
+
+size_t aspa_vector_find_cas(struct aspa_vector *vector, uint32_t customer_as)
+{
+	// if the vector is empty we return an error
+	if (vector->size == 0 || vector->capacity == 0) {
+		return -1;
+	}
+
+	// left and right bound of our search space
+	size_t left = 0;
+	size_t right = vector->size;
+
+	// we stop if right and left crossed
+	while (left <= right) {
+		// current center
+		size_t center = left + ((right - left) / 2);
+		uint32_t center_value = vector->data[center].customer_as_number;
+
+		if (center_value == customer_as) {
+			// success found the value
+			return center;
+		} else if (center_value > customer_as) {
+			// value should be on the left site
+			right = center - 1;
+		} else {
+			// value should be on the right site
+			left = center + 1;
+		}
+	}
+
+	// element not found
+	return -1;
+}
+
+int aspa_vector_delete_at(struct aspa_vector *vector, size_t index)
+{
+	if (vector->size < index || vector->size == 0) {
```

Comment:
```suggestion
	if (vector->size <= index || vector->size == 0) {
```
e.g. if vector has size 3, one cannot delete the fourth element with index 3